### ISSUE TITLE
How to add variable to existing cube?

### ISSUE THREAD

**Issue by bjoern234** [JuliaDataCubes/YAXArrays.jl#65]
Hi! 
I have a cube e.g. with axes X,Y,Time, and Bands. Bands contains "NIR" and "RED" variables. How would I add a "NDVI" variable to Bands for each time step? Is that possible or do I have to compute a new "ndvi-cube" and combine or merge them? 

Thanks in advance
Bjoern

**Comment by felixcremer** [comment]
You have to compute a new ndvicube, but this is all done lazily and you don't have to save it to disk. 
```julia
julia> arr = rand(300,200,90, 4)

julia> x = RangeAxis(:X, 1:300)
X                   Axis with 300 Elements from 1 to 300

julia> y = RangeAxis(:Y, 1:200)
Y                   Axis with 200 Elements from 1 to 200

julia> t = RangeAxis(:time, 1:90)
time                Axis with 90 Elements from 1 to 90


julia> b = CategoricalAxis("Bands", ["Red", "Green", "Blue", "NIR"])
Bands               Axis with 4 elements: Red Green Blue NIR 

julia> yax = YAXArray([x,y,t,b], arr)
YAXArray with the following dimensions
X                   Axis with 300 Elements from 1 to 300
Y                   Axis with 200 Elements from 1 to 200
time                Axis with 90 Elements from 1 to 90
Bands               Axis with 4 elements: Red Green Blue NIR 
Total size: 164.79 MB

julia> ndvi = map((r,nir) ->(r-nir)/(r+nir),yax[Bands="Red"], yax[Bands="NIR"])
```

If you would like to combine the datasets again into a cube together you would have to split the cube and concatenate it again. 
```
julia> catax = CategoricalAxis("Bands", ["Red", "Green", "Blue", "NIR", "NDVI"])

julia> ndvicube = concatenatecubes([yax[Bands="Red"], yax[Bands="Green"], yax[Bands="Blue"], yax[Bands="NIR"], ndvi], catax)
YAXArray with the following dimensions
X                   Axis with 300 Elements from 1 to 300
Y                   Axis with 200 Elements from 1 to 200
time                Axis with 90 Elements from 1 to 90
Bands               Axis with 5 elements: Red Green Blue NIR NDVI 
Total size: 205.99 MB
```

**Comment by MarkusZehner** [comment]
Hi, 
how would i include a single variable (say orbit no. or ascending/descending, one value, comparable to the time axis, but not distinct values) into the cube?

As in the example below, add the dir axis into the cube:
```
julia> caxes(cube) 
3-element Vector{CubeAxis}:
 Y                   Axis with 22500 Elements from -855560.0 to -405580.0
 X                   Axis with 15000 Elements from -253755.0 to -553735.0
 Time                Axis with 1244 Elements from 2017-01-01T16:59:45 to 2019-12-31T05:33:44

julia> ax_dir
dir                 Axis with 1244 elements: A A .. A D 
```

Thanks!
Markus

**Comment by meggart** [comment]
Hi, thanks for the report. What you describe seems to be a Zarr error and should be fixed in Zarr.jl. Could you post the contents of the file "cube_from_above.zarr/sat/.zarray" and "cube_from_above.zarr/dir/.zarray"? We support several string data types in Zarr.jl, but there seem to be too many of them in numcodecs.

**Comment by MarkusZehner** [comment]
cube_from_above.zarr/sat/.zarray:
```
{
    "chunks": [
        714
    ],
    "compressor": {
        "blocksize": 0,
        "clevel": 5,
        "cname": "lz4",
        "id": "blosc",
        "shuffle": 1
    },
    "dtype": "|O",
    "fill_value": null,
    "filters": [
        {
            "id": "vlen-utf8"
        }
    ],
    "order": "C",
    "shape": [
        714
    ],
    "zarr_format": 2
}

```
cube_from_above.zarr/dir/.zarray:
```
{
    "chunks": [
        714
    ],
    "compressor": {
        "blocksize": 0,
        "clevel": 5,
        "cname": "lz4",
        "id": "blosc",
        "shuffle": 1
    },
    "dtype": "|O",
    "fill_value": null,
    "filters": [
        {
            "id": "vlen-utf8"
        }
    ],
    "order": "C",
    "shape": [
        714
    ],
    "zarr_format": 2
}
```

**Comment by MarkusZehner** [comment]
just an update: `push!(yax.properties, "orb"=>y.orb)` works fine with storing dataarrays within properties, credits to @felixcremer.

However, when saving, those properties are not kept:
```
julia> yax.properties
Dict{String, Any} with 2 entries:
  "sat_bin" => YAXArray with the following dimensions…
  "orb"     => YAXArray with the following dimensions…

julia> test = savecube(yax, "testcube_delete")
YAXArray with the following dimensions
Y                   Axis with 2 Elements from -667240.0 to -667220.0
X                   Axis with 2 Elements from -391375.0 to -391395.0
Time                Axis with 714 Elements from 2017-01-01T16:59:45 to 2019-12-31T05:33:44
TSdecomp            Axis with 8 elements: imf_1 imf_2 imf_3 imf_4 imf_5 imf_6 imf_7 raw 
Total size: 178.5 KB


julia> test
YAXArray with the following dimensions
Y                   Axis with 2 Elements from -667240.0 to -667220.0
X                   Axis with 2 Elements from -391375.0 to -391395.0
Time                Axis with 714 Elements from 2017-01-01T16:59:45 to 2019-12-31T05:33:44
TSdecomp            Axis with 8 elements: imf_1 imf_2 imf_3 imf_4 imf_5 imf_6 imf_7 raw 
Total size: 178.5 KB


julia> test.properties
Dict{String, Any}()
```

**Comment by felixcremer** [comment]
The saving of the properties will be fixed by #117. Is there anything else, that should be done here?

**Comment by lazarusA** [comment]
see https://juliadatacubes.github.io/YAXArrays.jl/dev/UserGuide/write.html#Append-to-a-Dataset for a cleaner way when it comes to append new variables.
